core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -0,0 +1,222 @@
+      }
+      .map(ContainerId.apply)
+  }
+
+  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress] = {
+    Future {
+      blocking {
+        val pod =
+          kubeRestClient.pods().withName(id.asString).waitUntilReady(timeouts.inspect.length, timeouts.inspect.unit)
+        ContainerAddress(pod.getStatus().getPodIP())
+      }
+    }.recoverWith {
+      case e =>
+        log.error(this, s"Failed to get IP of Pod '${id.asString}' within timeout: ${e.getClass} - ${e.getMessage}")
+        Future.failed(new Exception(s"Failed to get IP of Pod '${id.asString}'"))
+    }
+  }
+
+  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", id.asString), timeouts.rm).map(_ => ())
+
+  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), timeouts.rm).map(_ => ())
+
+  def logs(id: ContainerId, sinceTime: String)(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    Source.fromFuture(Future {
+      blocking {
+        val sinceTimePart = if (!sinceTime.isEmpty) {
+          s"&sinceTime=${sinceTime}"
+        } else ""
+        val url = new URL(
+          URLUtils.join(
+            kubeRestClient.getMasterUrl.toString,
+            "api",
+            "v1",
+            "namespaces",
+            kubeRestClient.getNamespace,
+            "pods",
+            id.asString,
+            "log?timestamps=true" ++ sinceTimePart))
+        val request = new Request.Builder().get().url(url).build
+
+        val response = kubeRestClient.getHttpClient.newCall(request).execute
+        if (!response.isSuccessful) {
+          Future.failed(
+            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
+        }
+        response.body.string
+      }
+    }.map { output =>
+      val original = output.lines.toSeq
+      val relevant = original.dropWhile(s => !s.startsWith(sinceTime))
+      val result =
+        if (!relevant.isEmpty && original.size > relevant.size) {
+          // drop matching timestamp from previous activation
+          relevant.drop(1)
+        } else {
+          original
+        }
+      // map the logs to the docker json file format expected by
+      // ActionLogDriver.processJsonDriverLogContents
+      // TODO - jcrossley - you'll probably want to clean up / reimplement the logic intended by this with the new format
+      //var sentinelSeen = false
+      ByteString(
+        result
+          .map { line =>
+            val pos = line.indexOf(" ")
+            val ts = line.substring(0, pos)
+            val msg = line.substring(pos + 1)
+            // TODO - can we get the proper stream name from kubernetes? Some stuff is stderr
+            val stream = "stdout"
+            JsObject("log" -> msg.toJson, "stream" -> stream.toJson, "time" -> ts.toJson)
```

Comment:
What about using `LogLine(msg, stream, ts).toJson.compactPrint` instead? That would ensure the proper format and fieldnames.

Reply:
I don't seem to have those methods on `LogLine`. Is it enough to just `LogLine(msg, stream, ts)`?

Reply:
`LogLine` is a case-class, you'll likely need to import it. But then, yes, it should just be that.

Reply:
One of us may be confused. Just to make sure you know what this code is doing: we are adapting the k8s log output, making it look like the docker log output, so that we can re-use the docker classes in `KubernetesContainer.logs`. I concede it's a hack, but I think it means we can't use `LogLine` at this point.

Reply:
Right I'm aware. Using `LogLine` just **ensures** you're adapting correctly (because it's used for deserialization again). Both you're any my line should yield the same result, but using `LogLine` ensures you're using the right field names etc.